seastar-macros/src/lib.rs

```diff
@@ -0,0 +1,56 @@
+#[proc_macro_attribute]
+#[cfg(not(test))] // Work around for rust-lang/rust#62127
+pub fn main(
+    args: proc_macro::TokenStream,
+    item: proc_macro::TokenStream,
+) -> proc_macro::TokenStream {
+    let input = syn::parse_macro_input!(item as syn::ItemFn);
+    let args = syn::parse_macro_input!(args as syn::AttributeArgs);
+
+    let inputs = &input.sig.inputs;
+    let name = &input.sig.ident;
+    let body = &input.block;
+    let attrs = &input.attrs;
+
+    if input.sig.asyncness.is_none() {
+        let msg = "the async keyword is missing from the function declaration";
+        return syn::Error::new_spanned(input.sig.fn_token, msg)
+            .to_compile_error()
+            .into();
+    }
+
+    if name != "main" {
+        let msg = "only the main function is allowed to use #[seastar::main]";
+        return syn::Error::new_spanned(input.sig.inputs, msg)
+            .to_compile_error()
+            .into();
+    }
+
+    if !inputs.is_empty() {
+        let msg = "the main function cannot accept arguments";
+        return syn::Error::new_spanned(input.sig.inputs, msg)
+            .to_compile_error()
+            .into();
+    }
+
+    if !args.is_empty() {
+        let msg = "arguments for #[seastar::main] are not supported yet";
+        return syn::Error::new_spanned(input.sig.inputs, msg)
+            .to_compile_error()
+            .into();
+    }
+
+    let output = quote::quote! {
+        #(#attrs)*
+        fn main() {
```

Comment:
It would be really nice if this macro supported `main` functions that don't return `()`. A main function can return anything that implements the `Termination` trait: https://doc.rust-lang.org/stable/std/process/trait.Termination.html.

Perhaps you could:

- run the main function code,
- pass the return value through a mutex,
- unpack the mutex, call `.report()` and return the exit code?

Reply:
Good idea, I didn't know of that trait. This shouldn't be too hard to implement.